File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
import sys
from collections import deque
import math
import heapq
import random
import itertools
ipt = sys.stdin.readline
def minput(): return map(int, ipt().split())


def bfs(grid, visited):
    temp = 0
    while q:
        for _ in range(len(q)):
            i = q.popleft()
            if i<=50000 and not visited[2*i]:
                if grid[2*i] == 998244353:
                    return grid[i]
                q.append(2*i)
                grid[2*i]=grid[i]
                visited[2*i] = True
            if i>0 and not visited[i-1]:
                if grid[i-1] == 998244353:
                    return grid[i]+1
                q.append(i-1)
                grid[i-1]=grid[i]+1
                visited[i-1] = True
            if i<100000 and not visited[i+1]:
                if grid[i+1] == 998244353:
                    return grid[i]+1
                q.append(i+1)
                grid[i+1]=grid[i]+1
                visited[i+1] = True
 

 
n, k = minput()
if n == k:
    print(0)
    sys.exit()
grid = [0]*100001
grid[n], grid[k] = 0, 998244353
visited = [False]*100001
visited[n] = True
q = deque([n])
print(bfs(grid, visited))

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.cc
#include <bits/stdc++.h>
using namespace std;
const int INF=99824435;

int bfs(int grid[], bool visited[], queue<int>& q){
    int temp=0;
    while (int s=q.size()){
        for (int i=0;i<s;i++){
            int now = q.front(); q.pop();
            if (now<=50000 and not visited[2*now]){
                if (grid[2*now]>=INF) return grid[now];
                q.push(2*now);
                grid[2*now]=grid[now];
                visited[2*now]=true;
            } 
            if (now>0 and not visited[now-1]){
                if (grid[now-1]>=INF) return grid[now]+1;
                q.push(now-1);
                grid[now-1]=grid[now]+1;
                visited[now-1]=true;
            }
            if (now<100000 and not visited[now+1]){
                if (grid[now+1]>=INF) return grid[now]+1;
                q.push(now+1);
                grid[now+1]=grid[now]+1;
                visited[now+1]=true;
            }

        }
    }
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int n,k;
    cin>>n>>k;
    if (n==k){
        cout<<0;
        exit(0);
    }
    int grid[100001]={0,};
    grid[n]=0;grid[k]=INF;
    bool visited[100001]={false,};
    visited[n]=true;
    queue<int> q;
    q.push(n);
    cout<<bfs(grid, visited, q);
}